### PR TITLE
add policy for CIS-1.1 - 4.03 Ensure "Block Project-wide SSH keys" is enabled for VM instances

### PR DESCRIPTION
**Pull request by palani-ram-google-partner**
add policy for CIS-1.1 - 4.03 Ensure "Block Project-wide SSH keys" is enabled for VM instances

### CODE REVIEW

**Comment by palani-ram-google-partner** [comment]
@morgante , Please review this PR and let me know if there is any issue or good to merge.

File: policies/templates/gcp_compute_block_ssh_keys_v1.yaml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+kind: ConstraintTemplate
+metadata:
+  name: gcp-compute-block-ssh-keys-v1
+spec:
+  crd:
+    spec:
+      names:
+        kind: GCPComputeBlockSSHKeysConstraintV1
+      validation:
+        openAPIV3Schema:
```
Comment (morgante) [inline policies/templates/gcp_compute_block_ssh_keys_v1.yaml:25]: We should still specify a schema, but with no properties. cf. https://github.com/GoogleCloudPlatform/policy-library/blob/master/policies/templates/gcp_gke_dashboard_v1.yaml#L28

File: policies/templates/gcp_compute_block_ssh_keys_v1.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        
+        	# check if key and values are as expected
+        	metadata_config.items[i].key == "block-project-ssh-keys"
+        	metadata_config.items[i].value != "true"
+        
+        	message := sprintf("%v Block project-wide SSH keys is not enabled.", [asset.name])
+        	ancestry_path = lib.get_default(asset, "ancestry_path", "")
+        	metadata := {"ancestry_path": ancestry_path}
+        }
+        #ENDINLINE
```
Comment (morgante) [inline policies/templates/gcp_compute_block_ssh_keys_v1.yaml:68]: Nit: Please include a newline at the end of files.

File: samples/compute_block_ssh_keys.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+apiVersion: constraints.gatekeeper.sh/v1alpha1
+kind: GCPComputeBlockSSHKeysConstraintV1
+metadata:
+  name: compute_block_ssh_keys
+  annotations:
+    description: CIS-1.1 - 4.03 Ensure "Block Project-wide SSH keys" is enabled for VM instances
```
Comment (morgante) [inline samples/compute_block_ssh_keys.yaml:20]: CIS constraints should follow the existing annotation standard.

https://github.com/GoogleCloudPlatform/policy-library/blob/master/samples/compute_forbid_ip_forward.yaml#L20

File: validator/compute_block_ssh_keys.rego [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	asset := input.asset
+	asset.asset_type == "compute.googleapis.com/Instance"
+	instance := asset.resource.data
+	metadata_config := lib.get_default(instance, "metadata", {})
+
+	# check if key and values are as expected
+	metadata_config.items[i].key == "block-project-ssh-keys"
```
Comment (morgante) [inline validator/compute_block_ssh_keys.rego:31]: Unfortunately the logic for this isn't quite right. If `block-project-ssh-keys` isn't set at all, this `deny` condition will never occur.

What you probably need to do is add an additional function like this:

```
default metadata_blocks_project_keys(meta) = false
metadata_blocks_project_keys(meta) = true {
  metadatum := meta.items[_]
  metadatum.key == "block-project-ssh-keys"
  metadatum.value == "true"
}
```

Then within the core `deny` block you just call it:
```
  not metadata_blocks_project_keys(metadata_config)
```

File: validator/compute_block_ssh_keys.rego [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+deny[{
+	"msg": message,
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	asset := input.asset
+	asset.asset_type == "compute.googleapis.com/Instance"
+	instance := asset.resource.data
+	metadata_config := lib.get_default(instance, "metadata", {})
```
Comment (morgante) [inline validator/compute_block_ssh_keys.rego:28]: You should include an empty items list by default.

```suggestion
	metadata_config := lib.get_default(instance, "metadata", {"items": []})
```

File: validator/compute_block_ssh_keys_test.rego [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+package templates.gcp.GCPComputeBlockSSHKeysConstraintV1
+
+import data.test.fixtures.compute_block_ssh_keys.assets as fixture_instances
+import data.test.fixtures.compute_block_ssh_keys.constraints as fixture_constraints
+
+# Find all violations on our test cases
+find_violations[violation] {
```
Comment (morgante) [inline validator/compute_block_ssh_keys_test.rego:22]: Please use the new test utils. Example: https://github.com/GoogleCloudPlatform/policy-library/blob/master/validator/enforce_labels_test.rego#L19-L20

File: validator/compute_block_ssh_keys_test.rego [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	instance := data.instances[_]
+	constraint := data.test_constraints[_]
+	issues := deny with input.asset as instance
+		 with input.constraint as constraint
+
+	total_issues := count(issues)
+	violation := issues[_]
+}
+
+# Confim no violations with no instances
```
Comment (morgante) [inline validator/compute_block_ssh_keys_test.rego:32]: For test cases, please test the following scenarios:

1. No instances at all
2. One instance with the correct key, one without
3. No instances with the correct key
4. An instance without metadata configured at all (`metadata_config` doesn't exist).

File: validator/test/fixtures/compute_block_ssh_keys/constraints/block_ssh_keys_default/data.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+#
+apiVersion: constraints.gatekeeper.sh/v1alpha1
+kind: GCPComputeBlockSSHKeysConstraintV1
+metadata:
+  name: compute-block-ssh-keys-v1
+  annotations:
+    description: Ensure "Block Project-wide SSH keys" is enabled for VM instances
+spec:
+  severity: high
+  parameters: { }
```
Comment (morgante) [inline validator/test/fixtures/compute_block_ssh_keys/constraints/block_ssh_keys_default/data.yaml:23]:
```suggestion
  parameters: {}
```

**Review by morgante — CHANGES_REQUESTED** [review]
Thanks for contributing.

**Comment by palani-ram-google-partner** [comment]
@morgante 
We can close this PR. We do not need to merge this PR.

Created a new PR with these review comments fixed.
https://github.com/GoogleCloudPlatform/policy-library/pull/405

**Pull request closed without merging**